crates/forge/src/sierra_casm_runner.rs

```diff
@@ -0,0 +1,558 @@
+    pub fn create_code_footer() -> Vec<Instruction> {
+        casm! {
+            // Add a `ret` instruction used in libfuncs that retrieve the current value of the `fp`
+            // and `pc` registers.
+            ret;
+        }
+        .instructions
+    }
+
+    pub fn get_casm_program(&self) -> &CairoProgram {
+        &self.casm_program
+    }
+}
+
+fn create_metadata(
+    sierra_program: &cairo_lang_sierra::program::Program,
+    metadata_config: Option<MetadataComputationConfig>,
+) -> Result<Metadata, RunnerError> {
+    if let Some(metadata_config) = metadata_config {
+        calc_metadata(sierra_program, metadata_config, false).map_err(|err| match err {
+            MetadataError::ApChangeError(err) => RunnerError::ApChangeError(err),
+            MetadataError::CostError(_) => RunnerError::FailedGasCalculation,
+        })
+    } else {
+        Ok(Metadata {
+            ap_change_info: calc_ap_changes(sierra_program, |_, _| 0)?,
+            gas_info: GasInfo {
+                variable_values: Default::default(),
+                function_costs: Default::default(),
+            },
+        })
+    }
+}
+
+// casm_run::run_function
+pub fn run_function<'a, 'b: 'a, Instructions>(
+    vm: &mut VirtualMachine,
+    instructions: Instructions,
+    builtins: Vec<BuiltinName>,
+    additional_initialization: fn(
+        context: RunFunctionContext<'_>,
+    ) -> Result<(), Box<CairoRunError>>,
+    hint_processor: &mut TestExecutionSyscallHandler,
+    hints_dict: HashMap<usize, Vec<HintParams>>,
+) -> Result<(Vec<Option<Felt252>>, usize), Box<CairoRunError>>
+where
+    Instructions: Iterator<Item = &'a Instruction> + Clone,
+{
+    let data: Vec<MaybeRelocatable> = instructions
+        .flat_map(|inst| inst.assemble().encode())
+        .map(Felt252::from)
+        .map(MaybeRelocatable::from)
+        .collect();
+
+    let data_len = data.len();
+    let program = Program::new(
+        builtins,
+        data,
+        Some(0),
+        hints_dict,
+        ReferenceManager {
+            references: Vec::new(),
+        },
+        HashMap::new(),
+        vec![],
+        None,
+    )
+    .map_err(CairoRunError::from)?;
+    let mut runner = CairoRunner::new(&program, "all_cairo", false)
+        .map_err(CairoRunError::from)
+        .map_err(Box::new)?;
+
+    let end = runner.initialize(vm).map_err(CairoRunError::from)?;
+
+    additional_initialization(RunFunctionContext { vm, data_len })?;
+
+    runner
+        .run_until_pc(end, vm, hint_processor)
+        .map_err(CairoRunError::from)?;
+    runner
+        .end_run(true, false, vm, hint_processor)
+        .map_err(CairoRunError::from)?;
+    runner.relocate(vm, true).map_err(CairoRunError::from)?;
+
+    // changed region
+    finalize(
+        vm,
+        &runner,
+        &mut hint_processor
+            .contract_execution_syscall_handler
+            .cheatable_syscall_handler
+            .syscall_handler,
+        0,
+        2,
```

Comment:
Not sure about these two tbh 😕 
In the cheatnet we have those values prepared before:
- `n_total_args` is the length of args passed to call (what value should be passed here?)
- `program_extra_data_length` is always set to `2` in blockifier

Reply:
I've checked and neither the first nor the second have an influence on gas 🤷